EIPS/eip-5805.md

```diff
@@ -0,0 +1,401 @@
+        chainid,
+        address(this)
+));
+```
+
+In other words, the message is the EIP-712 typed structure:
+
+```js
+{
+  "types": {
+    "EIP712Domain": [
+      {
+        "name": "name",
+        "type": "string"
+      },
+      {
+        "name": "version",
+        "type": "string"
+      },
+      {
+        "name": "chainId",
+        "type": "uint256"
+      },
+      {
+        "name": "verifyingContract",
+        "type": "address"
+      }
+    ],
+    "Delegation": [{
+      "name": "delegatee",
+      "type": "address"
+      },
+      {
+        "name": "nonce",
+        "type": "uint256"
+      },
+      {
+        "name": "expiry",
+        "type": "uint256"
+      }
+    ],
+    "primaryType": "Permit",
+    "domain": {
+      "name": contractName,
+      "version": version,
+      "chainId": chainid,
+      "verifyingContract": contractAddress
+  },
+  "message": {
+    "delegatee": delegatee,
+    "nonce": nonce,
+    "expiry": expiry
+  }
+}}
+```
+
+Note that nowhere in this definition do we refer to `msg.sender`. The caller of the `delegateBySig` function can be any address.
+
+When this function is successfully executed, the delegator's nonce MUST be incremented to prevent replay attacks.
+
+#### nonces
+
+This function returns the current nonce for a given account.
+
+Signed delegations (see `delegateBySig`) are only accepted if the nonce used in the EIP-712 signature matches the return of this function. This value of `nonce(delegator)` should be incremented whenever a call to `delegateBySig` is performed on behalf of `delegator`.
+
+This function MUST be implemented
+
+```yaml
+- name: nonces
+  type: function
+  stateMutability: view
+  inputs:
+    - name: account
+      type: delegator
+  outputs:
+    - name: nonce
+      type: uint256
+```
+
+### Events
+
+#### DelegateChanged
+
+`delegator` changes the delegation of its assets from `fromDelegate` to `toDelegate`.
+
+MUST be emitted when the delegate for an account is modified by `delegate(address)` or `delegateBySig(address,uint256,uint256,uint8,bytes32,bytes32)`.
+
+```yaml
+- name: DelegateChanged
+  type: event
+  inputs:
+    - name: delegator
+      indexed: true
+      type: address
+    - name: fromDelegate
+      indexed: true
+      type: address
+    - name: toDelegate
+      indexed: true
+      type: address
+```
+
+#### DelegateVotesChanged
+
+`delegate` available voting power changes from `previousBalance` to `newBalance`.
+
+This MUST be emitted when:
+- an account (that holds more than 0 assets) updates its delegation from or to `delegate`,
+- an asset transfer from or to an account that is delegated to `delegate`.
+
+```yaml
+- name: DelegateVotesChanged
+  type: event
+  inputs:
+    - name: delegate
+      indexed: true
+      type: address
+    - name: previousBalance
+      indexed: false
+      type: uint256
+    - name: newBalance
+      indexed: false
+      type: uint256
+```
+
+### Solidity interface
+
+```sol
+interface IERC_XXXX {
```

Comment:
[Optional] Do you want to consider adding EIP-165 identifer?

Reply:
I'm not sure, that is a good point.

I love EIP-165 but the truth is I don't see it used that often in the field.

Reply:
also, with "now()" being optional, I'm not sure what the identifier should be?

Reply:
> also, with "now()" being optional, I'm not sure what the identifier should be?

Good point. It also makes restriction usage of EIP-165. I think the common approach is to make EIP-165 out of all required field, and then any optional one will have an `IERC5805Extension` kind of thing which has its own EIP-165 identifier.